### PR TITLE
Add possibility to use S3 job folder as stage in/out area for the job

### PR DESCRIPTION
**Pull request by enrico-usai**
+ At submission time an S3 folder is created for the job,
  containing the command file (if there) and the environment (if there).
+ The s3 bucket folder is available in the `awsbstat -d` command.
+ We are adding `PutObject` permission to the Job Role.

+ We are exporting in the job environment the `PCLUSTER_JOB_S3_URL`, to be used for stage-in/out by the job.
+ Added `--input-file file1 --input-file file2` option to the awsbsub to stage-in input files from the client.
+ Added `--working-dir` to give the user the possibility to specify the working-directory for the job in the compute instance
+ Added `--parent-working-dir` to give the user the possibility to specify the parent-working-directory for the job in the compute instance it is an alternative to the previous

Examples:
```
$ awsbsub -cf test.sh --> pwd: /job-08f80690-947d-40e5-9bc7-5ad78b729991

$ awsbsub -pw /custom/ -cf test.sh --> pwd: /custom/job-cfc1faa6-554a-46d9-84a2-5a2c16c8297a

$ awsbsub -w /customwd/ -cf test.sh --> pwd: /customwd
```

### CODE REVIEW

File: cloudformation/batch-substack.cfn.json [hunk 1]
```diff
@@ -120,7 +120,7 @@
           ]
         },
         "ManagedPolicyArns": [
-          "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess",
+          "arn:aws:iam::aws:policy/AmazonS3FullAccess",
```
Comment (mauri-melato) [inline cloudformation/batch-substack.cfn.json]: Isn't it too much?

File: cli/awsbatch/utils.py [hunk 2]
```diff
@@ -99,3 +99,36 @@ def is_job_array(job):
...
+
+        :param boto3_factory: initialized Boto3ClientFactory object
+        :param s3_bucket: S3 bucket to use
+        :param s3_bucket: S3 folder on which put the files (optional)
+        """
+        self.boto3_factory = boto3_factory
+        self.s3_bucket = s3_bucket
+        self.s3_folder = s3_folder
+
+    def put_file(self, file_path, key_name, timeout, s3_folder=None):
```
Comment (mauri-melato) [inline cli/awsbatch/utils.py]: Isn't ttl or time_to_live more appropriate in this context?

File: cli/awsbatch/utils.py [hunk 3]
```diff
@@ -99,3 +99,36 @@ def is_job_array(job):
...
+
+        :param file_path: file to upload
+        :param key_name: S3 key to create
+        :param timeout: S3 expiration time in seconds
+        :param s3_bucket: S3 folder on which put the files (optional, the default is the one passed to the constructor)
+        """
+        default_expiration = 30  # minutes
+        expires = datetime.datetime.now() + datetime.timedelta(minutes=default_expiration)
+        if timeout:
+            expires += datetime.timedelta(seconds=timeout)
```
Comment (mauri-melato) [inline cli/awsbatch/utils.py]: The expiration is going to be  `now + 30 min + timeout` that's not one would expect setting the timeout.

File: cli/awsbatch/utils.py [hunk 4]
```diff
@@ -99,3 +99,36 @@ def is_job_array(job):
...
+    def put_file(self, file_path, key_name, timeout, s3_folder=None):
+        """
+        Upload a file to an s3 bucket.
+
+        :param file_path: file to upload
+        :param key_name: S3 key to create
+        :param timeout: S3 expiration time in seconds
+        :param s3_bucket: S3 folder on which put the files (optional, the default is the one passed to the constructor)
+        """
+        default_expiration = 30  # minutes
```
Comment (mauri-melato) [inline cli/awsbatch/utils.py]: Why 30 mins?

File: cli/awsbatch/utils.py [hunk 5]
```diff
@@ -99,3 +99,36 @@ def is_job_array(job):
...
+        self.s3_bucket = s3_bucket
+        self.s3_folder = s3_folder
+
+    def put_file(self, file_path, key_name, timeout, s3_folder=None):
+        """
+        Upload a file to an s3 bucket.
+
+        :param file_path: file to upload
+        :param key_name: S3 key to create
+        :param timeout: S3 expiration time in seconds
```
Comment (mauri-melato) [inline cli/awsbatch/utils.py]: Is it really useful to have a fine granularity with seconds or accepting minutes would be just fine and more friendly for users?

File: cli/awsbatch/awsbsub.py [hunk 6]
```diff
@@ -164,82 +163,62 @@ def _validate_parameters(args):
...
             env_blacklist = args.env_blacklist if args.env_blacklist else config.env_blacklist
-            _get_env_and_upload(s3_uploader, args.env, env_blacklist, job_folder, env_file, args.timeout, log)
+            _get_env_and_upload(s3_uploader, args.env, env_blacklist, env_file, args.timeout, log)
 
         # upload job script
         if args.command_file:
             # existing script file
             try:
-                s3_uploader.put_file(args.command, job_folder + job_script, args.timeout)
+                s3_uploader.put_file(args.command, job_script, args.timeout)
```
Comment (mauri-melato) [inline cli/awsbatch/awsbsub.py]: Can we just have the ttl on the job_folder so all the stuff down there will inherit the same ttl ?

File: cloudformation/batch-substack.cfn.json (hunk 1 above)
Comment (sean-smith) [reply cloudformation/batch-substack.cfn.json]: Also the `arn:aws:` won't work in non-commercial partitions, I know batch isn't supported there yet but we should make changes as if it is. Please use `${AWS::Partition}`

File: cli/awsbatch/awsbsub.py (hunk 6 above)
Comment (enrico-usai) [reply cli/awsbatch/awsbsub.py]: We are removing the Expires property since doesn't work as expected.

File: cli/awsbatch/utils.py [hunk 7]
```diff
@@ -99,3 +99,42 @@ def is_job_array(job):
...
+        :param s3_bucket: S3 bucket to use
+        :param default_folder: S3 folder on which put the files (optional)
+        """
+        self.s3_client = boto3_factory.get_client("s3")
+        self.s3_bucket = s3_bucket
+        self.default_folder = default_folder
+        if default_folder:
+            self.create_folder(default_folder)
+
+    def create_folder(self, folder):
```
Comment (demartinofra) [inline cli/awsbatch/utils.py]: "private" method?

File: cli/awsbatch/awsbsub.py [hunk 8]
```diff
@@ -57,6 +56,14 @@ def _get_parser():
         help="Identifies that the command is a file to be transferred to the compute instances",
         action="store_true",
     )
+    parser.add_argument("-jwp", "--job-wd-parent", help="The parent folder of the job working directory")
```
[the remark this answers is not in the document]
Comment (enrico-usai) [reply cli/awsbatch/awsbsub.py]: Good catch! We are going to change it to be simply "working dir".

File: cli/awsbatch/utils.py (hunk 7 above)
Comment (enrico-usai) [reply cli/awsbatch/utils.py]: yep

File: cloudformation/batch-substack.cfn.json [hunk 9]
```diff
@@ -121,8 +121,29 @@
           ]
         },
         "ManagedPolicyArns": [
-          "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess",
+          {
+            "Fn::Sub": "arn:${AWS::Partition}:iam::aws:policy/AmazonS3ReadOnlyAccess"
+          },
           "arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy"
```
Comment (demartinofra) [inline cloudformation/batch-substack.cfn.json]: I know it's not part of your PR but I was just wondering if we should also use {AWS::Partition} here

File: cli/awsbatch/awsbsub.py [hunk 10]
```diff
@@ -57,6 +56,25 @@ def _get_parser():
...
+        "-w",
+        "--working-dir",
+        help="The folder to use as job working directory. "
+        "If not specified the job will be executed in the job-<AWS_BATCH_JOB_ID> subfolder of the user's home",
+    )
+    parser.add_argument(
+        "-pw",
+        "--parent-working-dir",
+        help="Parent folder for the job working directory. If not specified it is the user's home. "
+        "A subfolder named job-<AWS_BATCH_JOB_ID> will be created on it. Alternative to the --working-dir parameter",
```
Comment (demartinofra) [inline cli/awsbatch/awsbsub.py]: created in it.

File: cloudformation/batch-substack.cfn.json (hunk 9 above)
Comment (enrico-usai) [reply cloudformation/batch-substack.cfn.json]: fixed

File: cli/awsbatch/awsbsub.py (hunk 10 above)
Comment (enrico-usai) [reply cli/awsbatch/awsbsub.py]: fixed

**Pull request merged**